src/CTA.Rules.Update/CodeReplacer.cs

```diff
@@ -72,56 +115,63 @@ public CodeReplacer(List<SourceFileBuildResult> sourceFileBuildResults, ProjectC
+                            try
+                            {
+                                runResult = projectLevelAction.ProjectFileActionFunc(_projectConfiguration.ProjectPath,
+                                    projectType,
+                                    _projectConfiguration.TargetVersions,
+                                    projectActions.PackageActions.Distinct().ToDictionary(p => p.Name, p => p.Version),
+                                    projectActions.ProjectReferenceActions.ToList(),
+                                    _metadataReferences);
+                            }
+                            catch (Exception ex)
+                            {
+                                var actionExecutionException = new ActionExecutionException(projectLevelAction.Name, projectLevelAction.Key, ex);
+                                projectActionExecution.InvalidExecutions = 1;
+                                LogHelper.LogError(actionExecutionException);
+                            }
                         }
                     }
-                    else if (projectLevelAction.ProjectFileActionFunc != null)
+                    if (!string.IsNullOrEmpty(runResult))
                     {
-                        try
-                        {
-                            runResult = projectLevelAction.ProjectFileActionFunc(_projectConfiguration.ProjectPath, projectType, _projectConfiguration.TargetVersions, projectActions.PackageActions.Distinct().ToDictionary(p => p.Name, p => p.Version), projectActions.ProjectReferenceActions.ToList());
-                        }
-                        catch (Exception ex)
-                        {
-                            var actionExecutionException = new ActionExecutionException(projectLevelAction.Name, projectLevelAction.Key, ex);
-                            projectActionExecution.InvalidExecutions = 1;
-                            LogHelper.LogError(actionExecutionException);
-                        }
+                        projectActionExecution.Description = string.Concat(projectActionExecution.Description, ": ", runResult);
+                        projectRunActions.Add(projectActionExecution);
+                        LogHelper.LogInformation(projectLevelAction.Description);
                     }
                 }
-                if (!string.IsNullOrEmpty(runResult))
+
+                if (!actionsPerProject.TryAdd(Constants.Project, projectRunActions))
                 {
-                    projectActionExecution.Description = string.Concat(projectActionExecution.Description, ": ", runResult);
-                    projectRunActions.Add(projectActionExecution);
-                    LogHelper.LogInformation(projectLevelAction.Description);
+                    LogHelper.LogError(new FilePortingException(Constants.Project, new Exception("Error adding project to actions collection")));
                 }
             }
```

Comment:
Same as above- consider moving the logic in this `if` block to a private function called `ApplyProjectLevelActions` or something similar